src/util/math.js

```diff
@@ -87,16 +109,142 @@ export const calculateEnrichedSignatures = ({
+  return newData.orderedNodes().map((d) => d[idKey]);
+};
+
+// calculate data for volcano plot
+// adapted from https://github.com/greenelab/adage-server/blob/master/interface/src/app/analyze/analysis/sampleBin.js
+export const calculateVolcanoSignatures = ({
+  signatureList,
+  activities,
+  diamondGroup,
+  spadeGroup
+}) => {
+  // if we dont have all we need, exit
+  if (
+    !isArray(signatureList) ||
+    !signatureList.length ||
+    !isArray(activities) ||
+    !activities.length ||
+    !isArray(diamondGroup) ||
+    diamondGroup.length < 2 ||
+    !isArray(spadeGroup) ||
+    spadeGroup.length < 2
+  )
+    return [];
+
+  // take signature id and sample id, and find value of associated activity
+  const getActivity = (signatureId, sampleId) =>
+    (
+      activities.find(
+        (activity) =>
+          activity.signature === signatureId && activity.sample === sampleId
+      ) || {}
+    ).value;
+
+  // remove any signatures that are not part of activities
+  signatureList = signatureList.filter((signature) =>
+    activities.find((activity) => activity.signature === signature.id)
+  );
+
+  // for each signature
+  let volcanoSignatures = signatureList.map((signature) => {
+    // get associated activity for each diamond sample
+    const diamondActivities = diamondGroup
+      .map((sampleId) => getActivity(signature.id, sampleId))
+      .filter((activity) => activity);
+    // get associated activity for each spade sample
+    const spadeActivities = spadeGroup
+      .map((sampleId) => getActivity(signature.id, sampleId))
+      .filter((activity) => activity);
+
+    // compute difference between diamond and spade activity means
+    let meanDiff = mean(diamondActivities) - mean(spadeActivities);
+    meanDiff = meanDiff.toFixed(5);
+    // compute p value of signature based using t test from ttest library
+    const pValue = ttest(diamondActivities, spadeActivities);
+
+    return {
+      ...signature,
+      diamondActivities,
+      spadeActivities,
+      meanDiff,
+      pValue
+    };
+  });
+
+  // extract p values from volcano signatures into array
+  const pValues = volcanoSignatures.map((signature) => signature.pValue);
+  // correct p values using false discovery rate function from multtest library
+  const correctedPValues = fdr(pValues);
+  // put corrected p values back into volcano signatures
+  volcanoSignatures = volcanoSignatures.map((volcanoSignature, index) => ({
+    ...volcanoSignature,
+    // negative log 10 transform and round decimal point
+    pValue: -Math.log10(correctedPValues[index]).toFixed(5)
+  }));
+
+  return volcanoSignatures;
+};
+
+// perform two-sample, unpaired, welch's (student's) t-test and return p value
+// from https://github.com/jstat/jstat/issues/189#issuecomment-469255296
+// note: cannot use the 'ttest' library because downstream dependency is
+// written in web assembly and cannot be compiled by 'workerize-loader' yet
+const ttest = (array1, array2) => {
```

Comment:
The correctness of this function is critical, so we may need some tests to ensure that it will survive all edge cases. If you don't want to include the tests in this PR, can you add a new issue for now? Thanks.

Reply:
I agree. This, and pretty much everything else in `math.js` really need some rigorous testing and review from statistical people in the lab.

I think for now, I'll just spot check this against the previous interface, and then merge it. But I think the next big thing I want to do (perhaps before starting the Signatures page), it go back and get some real test coverage for the entire frontend. And that will include talking with people in the lab and getting some statistical test cases from them, comparing our code to other implementations (eg in python).

Reply:
:+1:

Reply:
Issue to track it:
https://github.com/greenelab/adage-frontend/issues/111